File: cloud-indexer/deps/dart-packages/modular_core/lib/graph/query/convert.dart
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:parser/expression.dart';

import 'query.dart';

/// Converts from the parse-tree [parser.PatternExpr] format into the modular
/// internal representation [GraphQuery].
GraphQuery patternExprToGraphQuery(final PatternExpr input) {
  final List<String> inEdgeLabels = new List<String>.from(
      input.property.labels.map((Label each) => each.uri.toString()));
  final List<String> valueLabels = new List<String>.from(
      input.property.representations.map((Label each) => each.uri.toString()));
  bool isRequired = true;
  bool isRepeated = false;
  if (input.property.cardinality.isOptional) isRequired = false;
  if (input.property.cardinality.isRepeated) isRepeated = true;

  return new GraphQuery(inEdgeLabels,
      isRequired: isRequired,
      isRepeated: isRepeated,
      valueLabels: valueLabels,
      childConstraints: input.children.map(patternExprToGraphQuery).toList());
}

/// Converts from the parse-tree [parser.PathExpr] format into the modular
/// internal representation [GraphQuery].
GraphQuery pathExprToGraphQuery(final PathExpr input) {
  final PatternExpr patternExpr = input.properties.reversed.fold(
      null,
      (final PatternExpr childExpr, final Property parentProperty) =>
          new PatternExpr(
              parentProperty, childExpr == null ? null : [childExpr]));

  return patternExprToGraphQuery(patternExpr);
}

File: benchmark/lib/convert.cc
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "apps/ledger/benchmark/lib/convert.h"

#include <string.h>

namespace benchmark {

std::string ToString(fidl::Array<uint8_t>& data) {
  std::string ret;
  ret.resize(data.size());
  memcpy(&ret[0], data.data(), data.size());
  return ret;
}

fidl::Array<uint8_t> ToArray(const std::string& val) {
  auto ret = fidl::Array<uint8_t>::New(val.size());
  memcpy(ret.data(), val.data(), val.size());
  return ret;
}

}  // namespace benchmark
